ETCD-318: Adding MicroShift etcd enhancement (#1242)
openshift/enhancements: enhancements/microshift/etcd-as-a-transient-systemd-unit.md

While this is not ideal, decoupling etcd into its own process to resolve the dependency conflict is the first step and getting the etcd Go build version aligned back to 1.16 can be a future goal.

##### RPM Pros
* This scenario should incur the least amount of non-etcd-related overhead since the binary is being run directly.
* Integration is straight-forward; the etcd rpm is another MicroShift dependency, embedded into the rpm-ostree and distributed the same way as other dependencies.
* Clear, consistent architecture: There is only either RPM-installed content or content hosted on the cluster - no third way of installing/running things.
    
##### RPM Cons
* New plumbing will need to be created to (or potentially existing will need to be updated/refurbished to) build etcd and package it into an RPM.
* etcd will need to be built with a newer version of Golang (1.19) than is currently expected from upstream (1.16); etcd will be built under the MicroShift build root, locking the Golang version.

### Workflow Description

> TODO

Reviewer: This section should eventually contain a description of how the user starts MicroShift and how etcd ends up running. If we want to describe the separate developer workflow, that could be included as a variation.

#### Variation [optional]

TODO

### API Extensions

> TODO

Reviewer: I think you can say "N/A" for this section, since we're not changing any APIs.

### Implementation Details/Notes/Constraints [optional]

> TODO

Reviewer: We have been discussing a multi-step approach, starting with the process management change, then adding etcd in a go submodule to the microshift repo, then setting up builds out of the etcd repo and introducing a separate RPM. The first 2 steps let us reach a supportable state, but I think we want that third step for GA, don't we?

We should write all of that down here in this section.d

### Risks and Mitigations

> TODO

Reviewer: One risk is that we could have an issue if the etcd process does not exit and we try to start it again. What are we doing to protect against that situation? Do we need to do something other than relying on systemd?

### Drawbacks

> TODO

Reviewer: We will have a more complicated build and an extra package to manage. We consider those acceptable tradeoffs to be able to deliver etcd fixes quickly and to decouple the dependencies and allow microshift to build.

## Design Details

### Open Questions

#### RPM Build Cadence
How do we get to pre-release version built out for every sprint?
* How do we get fast-fix changes to customers that need/want them before a proper release?

#### Naming
Do we want to have it be just `etcd` or `microshift-etcd` or other?

> * Do we want to somehow indicate that this is specifically for MicroShift and not for general use; does that matter?

Reviewer: What would happen if we just add an etcd binary to the microshift RPM, without creating a separate RPM for now? Then when we have RPM builds out of the openshift/etcd repo we could remove the binary from the microshift RPM and add a dependency on the other RPM at the same time.

Reviewer: We can remove this open question.

#### Versioning

* How should we version the etcd RPM; should we follow the OpenShift versioning for etcd.
  * New versioning scheme that maps to an OpenShift/Upstream version
  * Go with the etcd version: v3.5.4
    * How do we reflect a patch or cherrypick update from OpenShift/etcd?
  * OpenShift version: v4.12
    * What is the release cadence for the etcd RPM/MicroShift as a whole? v4.12.1, v4.12.2, etc.
    * How do we reflect an update to the etcd version for a particular OpenShift release branch e.g 4.12 gets updated from etcd 3.5.4 to 3.5.5?

### Test Plan

**Note:** *Section not required until targeted at a release.*

> TODO

Reviewer: We should do the standard end-to-end tests. Can we describe some tests for things like etcd dying while microshift is running and vice versa?

### Graduation Criteria

TODO

#### Dev Preview -> Tech Preview

TODO

#### Tech Preview -> GA

TODO

#### Removing a deprecated feature

TODO

### Upgrade / Downgrade Strategy

TODO

### Version Skew Strategy

> TODO

Reviewer: Can we enforce not supporting version skew if we're shipping microshift and etcd in separate RPMs?

### Operational Aspects of API Extensions

> TODO

Reviewer: N/A

#### Failure Modes

TODO

#### Support Procedures

> TODO

Reviewer: We should mention the log separation again here. Would it make sense to talk about backup and restore of the database or is that procedure the same as for OCP?

## Implementation History

TODO

> ## Alternatives

Reviewer 2: Another option would be to compile etcd into a [go plugin](https://pkg.go.dev/plugin) and then continue to execute it in a goroutine. This would allow you to have a completely separate build chain for the binary, while not changing how etcd is run today. If necessary you can embed the plugin into the main Microshift binary. The installer is taking this approach with Terraform, since the various Terraform providers for different clouds often have conflicting dependencies.

Reviewer 2: Doug pointed out that you can still only have one version of each dependency otherwise the symbols conflict. (Additionally, the plugin has to be built with the exact same runtime version.) This presumably works in the installer because only one plugin is loaded at a time, and the dependency conflicts are between different Terraform plugins not between the installer and the plugin. So this won't work, but perhaps it's worth documenting why in the alternatives section.

---
title: etcd-as-a-transient-systemd-unit
authors:
  - dusk125
  - hasbro17
reviewers:

>   - "@tjungblu, etcd Team"

Reviewer 3: /lgtm!

This results in etcd dependencies being coupled with MicroShift which makes it harder to maintain and upgrade etcd independently of the platform like it is done with a separate openshift/etcd repo in the Openshift Container Platform (OCP).

This enhancement outlines an update to MicroShift to change how etcd is bundled and deployed with MicroShift in order to minimize risk of shared dependencies between etcd and OpenShift/Kubernetes leading to hard-to-trace bugs.

Changing the deployment of etcd, outlined below, would alleviate concerns about building etcd with a version of shared dependencies - mainly grpc - that etcd is not currently built against, and therefore, not tested against.

Additionally moving the etcd deployment into its own process would make it easier to debug and process logs via journatlctl.

### User Stories

1. "As an etcd developer, I want to maintain etcd versions independently of microshift dependencies."

2. "As a MicroShift Device Administrator, I want to observe and debug the etcd server logs separately from other microshift processes."

### Goals

> - The etcd binary is built and shipped separately from the microshift binary to ease maintainence.

Reviewer 4: Sorry for late comment, but in this step we won't ship etcd binary separately from microshift's binary, right? They will be in the same rpm, installed to the same location.
```suggestion
- The etcd is built as a separate binary to ease dependency management and shipped together with MicroShift binary in single RPM to ease deployment process.
```

Author: Yes that's much more clear, thanks!

---

> title: etcd-as-a-transient-systemd-unit

Author: ack

Additionally moving the etcd deployment into its own process would make it easier to debug and process logs via journatlctl.

### User Stories

1. "As an etcd developer, I want to maintain etcd versions independently of microshift dependencies."

2. "As a MicroShift Device Administrator, I want to observe and debug the etcd server logs separately from other microshift processes."

### Goals

- The etcd binary is built and shipped separately from the microshift binary to ease maintainence.
- The etcd server logs are easier to observe.

### Non-Goals

- Certificate management for etcd client authentication needs to be updated regardless of the delivery/execution implementation so this will be out of scope for this enhancement.
- Creating an RPM build root specifically for MicroShift-etcd; this should be eventually needed so we're building etcd with the Golang version it's expecting (1.16), instead of the same that MicroShift is built with (1.19).
  - This enhancement discusses this further in the 'etcd Binary from an RPM' section.
  - This work should be done in the future, but is out of scope for this enhancement.

> ## Proposal

Author: Yes

    
#### etcd Binary from an RPM
This will lay the etcd binary down into the same directory as the MicroShift binary so that in a development environment a local build of etcd will be used.

The RPM will be built in the same build root as MicroShift as a go submodule. This will resolve the dependency issues, but will cause etcd to be built with an unexpected version of Golang - Upstream/OpenShift etcd expects Go1.16, MicroShift is built with Go1.19.
While this is not ideal, decoupling etcd into its own process to resolve the dependency conflict is the first step and getting the etcd Go build version aligned back to 1.16 can be a future goal.

##### RPM Pros
* This scenario should incur the least amount of non-etcd-related overhead since the binary is being run directly.
* Integration is straight-forward; the etcd rpm is another MicroShift dependency, embedded into the rpm-ostree and distributed the same way as other dependencies.
* Clear, consistent architecture: There is only either RPM-installed content or content hosted on the cluster - no third way of installing/running things.
    
##### RPM Cons

> * New plumbing will need to be created to (or potentially existing will need to be updated/refurbished to) build etcd and package it into an RPM.

Reviewer: I think we settled on adding it to the existing MicroShift build and RPM to eliminate the need for any extra build pipeline work, right?

Author: Correct, the separate RPM for etcd is outside of this enhancements scope

The execution lifetime of the etcd service will be tied to that of MicroShift, so it will be automatically stopped if MicroShift is stopped by the user `systemctl stop microshift` and if MicroShift has an unexpected shutdown.
The execution of etcd should be completely transparent to the user; they would be able to see it running under systemd `systemctl status microshift-etcd` and collect logs from it `journalctl -u microshift-etcd`.

#### Variation [optional]

For the developer who wishes to run and debug MicroShift, microshift-etcd will detect that MicroShift is being run locally (not from a systemd unit) and change its execution of etcd to a direct binary execution; it will expect to find the etcd binary in the same directory as the MicroShift binary. This will allow the developer to build and debug both MicroShift and etcd locally.

### API Extensions

N/A

### Implementation Details/Notes/Constraints [optional]

> We are discussing a multi-step approach, starting with the process management change (this enhancements change), then adding etcd, in a go submodule, to the MicroShift repository, then setting up builds out of the etcd repo and introducing a separate RPM. The first two steps let us reach a supportable state, but we will need to address step three for the GA release of MicroShift.

Reviewer: Why do we need to block GA on the separate RPM?

Another test could be if etcd has an ungraceful shutdown; do the same steps as above, but using microshift-etcd's pid. Etcd should automatically come back up via the same systemd unit.

### Graduation Criteria

TODO

#### Dev Preview -> Tech Preview

TODO

#### Tech Preview -> GA

TODO

#### Removing a deprecated feature

TODO

### Upgrade / Downgrade Strategy

TODO

### Version Skew Strategy

Currently, there is no need for a version skew strategy as etcd will be built and delivered as a part of the MicroShift RPM, so it will not be possible for the MicroShift and etcd versions to get out of sync.

This may need to be revisited if we build and deliver etcd in its own RPM, but that is out of scope for this enhancement.

### Operational Aspects of API Extensions

N/A

#### Failure Modes

TODO

#### Support Procedures

##### Reading etcd Logs
With this enhancement, etcd logs will no longer show up in the MicroShift log stream, they will be in their own systemd log stream. The user/support can get these logs with `journalctl -u microshift-etcd` or similar command used for MicroShift itself.

> ##### Backup and Restore

Reviewer: We have a separate ticket to work on this, so let's leave this description here but not worry about fleshing out any more of the details in this enhancement.

## Motivation

Currently MicroShift [bundles upstream etcd](https://github.com/openshift/microshift/blob/f7f2260e4fbff61654f478fc149cb7052261f87f/go.mod#L21) into its binary and runs it as a goroutine.
This results in etcd dependencies being coupled with MicroShift which makes it harder to maintain and upgrade etcd independently of the platform like it is done with a separate openshift/etcd repo in the Openshift Container Platform (OCP).

This enhancement outlines an update to MicroShift to change how etcd is bundled and deployed with MicroShift in order to minimize risk of shared dependencies between etcd and OpenShift/Kubernetes leading to hard-to-trace bugs.

Changing the deployment of etcd, outlined below, would alleviate concerns about building etcd with a version of shared dependencies - mainly grpc - that etcd is not currently built against, and therefore, not tested against.

Additionally moving the etcd deployment into its own process would make it easier to debug and process logs via journatlctl.

### User Stories

1. "As an etcd developer, I want to maintain etcd versions independently of microshift dependencies."

> 2. "As an admin, I want to observe and debug the etcd server logs separately from other microshift processes."

Reviewer: ```suggestion
2. "As a MicroShift Device Administrator, I want to observe and debug the etcd server logs separately from other microshift processes."
```

We're trying to use consistent role names across the MicroShift enhancements.

### Goals

- The etcd binary is built and shipped separately from the microshift binary to ease maintainence.
- The etcd server logs are easier to observe.

### Non-Goals

- Certificate management for etcd client authentication needs to be updated regardless of the delivery/execution implementation so this will be out of scope for this enhancement.
- Creating an RPM build root specifically for MicroShift-etcd; this should be eventually needed so we're building etcd with the Golang version it's expecting (1.16), instead of the same that MicroShift is built with (1.19).
  - This enhancement discusses this further in the 'etcd Binary from an RPM' section.
  - This work should be done in the future, but is out of scope for this enhancement.

> ## Proposal

Reviewer: Let's clean this section up so it only includes the final proposal. The rest of the approaches should move down to alternatives and include a summary of why we rejected them.

We are discussing a multi-step approach, starting with the process management change (this enhancements change), then adding etcd, in a go submodule, to the MicroShift repository. Once this is in place, we might move the etcd build out to its own RPM, but that is outside of the scope for this enhancement.
This stepped approach will help keep these fundamental changes more manageable for reviewers, and will help decrease the likelihood of disruption due to these changes.

### Risks and Mitigations

The etcd execution lifetime should be bound to that of MicroShift (through systemd 'BindsTo' property); however, it may be possible that MicroShift exits and etcd is left running.
If this occurs, the user can also bring down etcd with `systemctl stop microshift-etcd`; if we find instances of this happening, we can add a check to the microshift-etcd startup to shutdown a running instance of microshift-etcd, if it exists.

### Drawbacks

We will have a more complicated build and an extra package to manage. We consider those acceptable tradeoffs to be able to deliver etcd fixes quickly and to decouple the dependencies that are causing MicroShift build issues.

> ## Design Details

Author: Are you suggesting in the Drawbacks section, or the Design Details section?

> ## Design Details

Reviewer: I think design details is more appropriate. It's a detail of the work we need to do.

> ## Design Details

Author: Sounds good, I just couldn't tell which code line was selected

### Open Questions

#### RPM Build Cadence
How do we get to pre-release version built out for every sprint?
* How do we get fast-fix changes to customers that need/want them before a proper release?

#### Naming
Do we want to have it be just `etcd` or `microshift-etcd` or other?
* Do we want to somehow indicate that this is specifically for MicroShift and not for general use; does that matter?

#### Versioning

* How should we version the etcd RPM; should we follow the OpenShift versioning for etcd.
  * New versioning scheme that maps to an OpenShift/Upstream version
  * Go with the etcd version: v3.5.4
    * How do we reflect a patch or cherrypick update from OpenShift/etcd?
  * OpenShift version: v4.12
    * What is the release cadence for the etcd RPM/MicroShift as a whole? v4.12.1, v4.12.2, etc.

>     * How do we reflect an update to the etcd version for a particular OpenShift release branch e.g 4.12 gets updated from etcd 3.5.4 to 3.5.5?

Reviewer: The RPM versioning question goes away, too, because we only have the 1 RPM.
